### ISSUE TITLE
Exporting when Tensor-Parallelism is >1 results in corrupted model

### ISSUE THREAD

**Issue by shaltielshmid** [NVIDIA/NeMo#14375]
**Describe the bug**

When I export a checkpoint from a train where I see TP > 1, then the resulting model is corrupted. 

**Steps/Code to reproduce bug**

Setup:

- 8xH100 on LambdaLabs
- Container: nvcr.io/nvidia/nemo:dev

Step 1: Train a model:

```python
from nemo.collections import llm
import nemo_run as run

recipe = llm.mistral_7b.finetune_recipe(dir="./checkpoints", num_nodes=1, num_gpus_per_node=8, peft_schema='none')
run.run(recipe, executor=run.LocalExecutor())
```

Step 2: After a checkpoint is saved:

*export.py*
```python
from nemo.collections import llm
llm.export_ckpt("./checkpoints/pt/checkpoints/....-last", "hf", "./hf_export/", overwrite=True)
```

Step 3: Try to use the model, and it outputs nonsense:
```python
from transformers import pipeline
import torch

oracle = pipeline('text-generation', './hf_export', device_map='cuda', torch_dtype=torch.bfloat16)
oracle('Hello my name is')
```

**Expected behavior**

The exported model should produce some reasonable completion to the text. 

**Environment overview (please complete the following information)**

 - Environment location: Docker
 - Method of install: `docker pull nvcr.io/nvidia/nemo:dev` && `docker run -it --rm --gpus all --network host --shm-size=8g -v $(pwd):/home/ubuntu -w /home/ubuntu nvcr.io/nvidia/nemo:dev /bin/bash`

**Comment by shaltielshmid** [comment]
Update: I am not replicating the bug when trying again with this minimal code - must be something else in my code, closing for now.

**Comment by shaltielshmid** [comment]
Update 2: I replicated the issue, and it only occurs when using TE+FP8 - I updated the minimal code in the original message.

**Comment by suiyoubi** [comment]
Hi @shaltielshmid , I wonder after completing step1 (with FP8+TE), are you able to successfully do inference in NeMo ? You can use the `llm.generate` API here: https://github.com/NVIDIA/NeMo/blob/15dc5cbe78f2d1830b3b6a88a0a9375deee3a025/nemo/collections/llm/api.py#L1037 

You can use similar script like this: 
```
import torch
import torch.distributed
from nemo.collections import llm
from megatron.core.inference.common_inference_params import CommonInferenceParams
import nemo.lightning as nl

if __name__ == "__main__":
    model_nemo_path = 'your/nemo/model/checkpoint'
    strategy = nl.MegatronStrategy(
        tensor_model_parallel_size=1,
        pipeline_model_parallel_size=1,
        context_parallel_size=1,
        sequence_parallel=False,
        setup_optimizers=False,
        store_optimizer_states=False,
    )
    trainer = nl.Trainer(
        accelerator="gpu",
        devices=1,
        num_nodes=1,
        strategy=strategy,
        plugins=nl.MegatronMixedPrecision(
            precision="bf16-mixed",
            params_dtype=torch.bfloat16,
            pipeline_dtype=torch.bfloat16,
            autocast_enabled=False,
            grad_reduce_in_fp32=False,
        ),
        callbacks=[]
    )

    prompts = ["Once upon a time,"]
    output = llm.generate(
        path=model_nemo_path,
        trainer=trainer,
        prompts=prompts,
        inference_params = CommonInferenceParams(num_tokens_to_generate=128),
        add_BOS=True,
    )
    print(output)
```
I want to understand if this is an exporting issue or training issue.

**Comment by shaltielshmid** [comment]
I believe the training works well - in my experiments the loss was the expected value and was learning, as expected. 

To be sure, I ran the inference code on the checkpoint, and the output makes sense:

```python
[InferenceRequest(request_id='0', prompt='Once upon a time,', sampling_params=SamplingParams(temperature=1.0, top_k=0, top_p=0.0, return_log_probs=False, return_segments=False, num_tokens_to_generate=128, top_n_logprobs=0), inference_parameters=None, prompt_tokens=[6481, 4482, 1032, 1495, 29493], arrival_time=1754423440.5360136, status=<Status.COMPLETED: 4>, encoder_prompt=None, generated_text=" I'd known very few Catholics. But locals filled every pew at weekday Mass.</s>", segments=None, generated_segments=None, generated_sequence_lengths=tensor([21], device='cuda:0', dtype=torch.int32), generated_tokens=tensor([ 1083, 29510, 29483,  3419,  1983,  2432,  9514,  1096,  1831, 29491,
         1860, 24005,  7542,  1780,  1525, 29495,  1206,  2587,  2234,  8344,
        29491], device='cuda:0'), generated_log_probs=None, generated_top_n_logprobs=None, generated_length=21)]
```

Might also be related to https://github.com/NVIDIA/NeMo/issues/14374

**Comment by shaltielshmid** [comment]
Update 3: The solution is already in the main branch:

https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/io/connector.py#L241-L242

I modified `/opt/NeMo/nemo/lightning/io/connector.py` in the container, and the export works. I'm using the nvcr.io/nvidia/nemo:dev container - is there an updated container? 

The docs recommend using this container, so I've been sticking with it.

**Comment by hemildesai** [comment]
Closing this issue since the solution is on `main`. Feel free to reopen if you are still facing it. You can find the latest nemo containers at https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo/tags
